ek9000App/src/devEK9000.cpp

```
@@ -91,33 +92,67 @@ void PollThreadFunc(void* param);
+					 epicsThreadGetStackSize(epicsThreadStackMedium),
+					 PollThreadFunc, NULL);
 }
 
+// This thread does two things:
+//      - Check the connection and reset the watchdog every other poll time.
+//      - Poll for new EL1xxx/EL3xxx/EL5xxx data every poll time.
 void PollThreadFunc(void*) {
+        int cnt = 0;
+	struct timeval start, finish;
+	double duration_ms;
 	while (true) {
+	        gettimeofday(&start, NULL);
 		// for (auto device : g_Devices) {
 		for (std::list<devEK9000*>::iterator it = g_Devices.begin(); it != g_Devices.end(); ++it) {
 			devEK9000* device = *it;
 			int status = device->Lock();
-			/* check connection */
-			bool connected = device->VerifyConnection();
-			if (!connected && device->m_connected) {
+			if (!cnt) {
+			    /* check connection every other loop */
+			    bool connected = device->VerifyConnection();
+			    if (!connected && device->m_connected) {
 				util::Warn("%s: Link status changed to DISCONNECTED\n", device->m_name);
 				device->m_connected = false;
-			}
-			if (connected && !device->m_connected) {
+			    }
+			    if (connected && !device->m_connected) {
 				util::Warn("%s: Link status changed to CONNECTED\n", device->m_name);
 				device->m_connected = true;
-			}
-			if (status)
+			    }
+			    if (status) {
+				device->Unlock();
 				continue;
-			uint16_t buf = 1;
-			device->m_driver->doModbusIO(0, MODBUS_WRITE_SINGLE_REGISTER, 0x1121, &buf, 1);
+			    }
+			    uint16_t buf = 1;
+			    if (device->m_driver->doModbusIO(0, MODBUS_WRITE_SINGLE_REGISTER, 0x1121, &buf, 1)) {
+				util::Error("%s: FAILED TO RESET WATCHDOG!\n", device->m_name);
+			    }
+			}
+			/* read EL1xxx/EL3xxx/EL5xxx data */
+			if (device->m_digital_cnt) {
+			    if (!device->m_driver->doModbusIO(0, MODBUS_READ_DISCRETE_INPUTS, 0,
+							      device->m_digital_buf, device->m_digital_cnt)) {
+				scanIoRequest(device->m_digital_io);
+			    } else
+				util::Error("%s: FAILED TO READ DIGITAL INPUTS!\n", device->m_name);
```

Hmm, I don't see how this failure to read can end up resulting in `READ/INVALID` being set on a per-record basis. 
Poll thread -> `m_digital_buf` -> `getEK9000IO` gets stale data

Am I misunderstanding the flow of data, or missing something else here?

---

We don't scan unless the read succeeds.  So if the read fails, the record just doesn't process.

So the data will be old, but so will the timestamp... the record doesn't update.

If we wanted to put out a read error, we could add {digital/analog}_failed flags here, scan in any case, and have getEK9000IO fail if the flag is set.

I'm open to suggestions on exactly what behavior we want here.
      1. Silent failure (keeping old data/timestamp without record processing).
      2. Setting a read error with the old data.
      3. Setting a read error with zero (or some other fixed value).

---

Ah, good point about the processing. I missed that important step.

So the current behavior is (1).

I personally think (2) is preferable - alarm sensitive clients will know there's something wrong, but non-alarm sensitive clients won't be thrown by a `0` value as could happen with (3). That is, there are so many scripts that are just `epics.caget(pvname)` and throw away alarm values entirely. A stale value is probably better than no value, in that case.

Other opinions, @JJL772 @ZLLentz @mamontironi?

---

I think you're right, actually.  Change pushed.

---

Changes look good. Thanks!